pkg/service/service.go

```diff
@@ -525,13 +536,50 @@ func publishIndexClaim(ctx context.Context, blobIndex blobindexlookup.BlobIndexL
 	)
 
 	digests := bytemap.NewByteMap[multihash.Multihash, struct{}](-1)
+	asyncDigests := bytemap.NewByteMap[multihash.Multihash, struct{}](-1)
+
+	// add priority digests for caching: root hash and shard hash(es).
+	digests.Set(link.ToCID(idx.Content()).Hash(), struct{}{})
+	for s := range idx.Shards().Iterator() {
+		digests.Set(s, struct{}{})
+	}
+
 	for _, slices := range idx.Shards().Iterator() {
 		for d := range slices.Iterator() {
-			digests.Set(d, struct{}{})
+			// add remaining hashes to the digests list up to the threshold, after
+			// which they are added to the other list for caching asynchronously.
+			if digests.Size() <= ProviderIndexAsyncCacheThreshold {
+				digests.Set(d, struct{}{})
+			} else {
+				asyncDigests.Set(d, struct{}{})
+			}
 		}
 	}
 
 	contextID := nb.Index.Binary()
+
+	var cacheErr error
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		err := provIndex.Cache(ctx, provider, contextID, digests.Keys(), meta, false)
+		errors.Join(cacheErr, err)
+		wg.Done()
+	}()
+	if asyncDigests.Size() > 0 {
+		wg.Add(1)
+		go func() {
+			err := provIndex.CacheAsync(ctx, provider, contextID, asyncDigests.Keys(), meta)
+			errors.Join(cacheErr, err)
+			wg.Done()
+		}()
+	}
```

Comment:
do we need to do these two in parallel or do we do it BECAUSE WE CAN? XD

Reply:
I figured that in the case where we actually do call `CacheAsync` we have a _lot_ of hashes and we don't want to be held up by both of these.

Reply:
I think it might be a bit of premature optimisation, but I don't think it hurts either, so 👍🏻